### PR TITLE
Fix magento codebase for Integration Tests to be compatible with PHP 8.1 part 2

### PR DESCRIPTION
**Pull request by SilinMykola**
### Description (*)
Fixed deprecations function php 8.1 to pass Integration Tests

### Related Issues (if relevant)
1. Fixes magento/magento2#34569

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by SilinMykola** [comment]
@magento run Static Tests

**Comment by SilinMykola** [comment]
@magento run Static Tests

**Comment by SilinMykola** [comment]
@magento run Static Tests

**Comment by SilinMykola** [comment]
@magento run Static Tests

**Comment by SilinMykola** [comment]
@magento run Static Tests

File: app/code/Magento/Catalog/Model/Indexer/Product/Flat/TableBuilder.php [hunk 1]
```diff
@@ -46,17 +48,20 @@ class TableBuilder
...
+        BuilderInterfaceFactory $tableBuilderFactory = null,
+        MetadataPool $metadataPool = null
     ) {
         $this->_productIndexerHelper = $productIndexerHelper;
         $this->resource = $resource;
         $this->_connection = $resource->getConnection();
-        $this->tableBuilderFactory = $tableBuilderFactory ?: \Magento\Framework\App\ObjectManager::getInstance()
+        $this->tableBuilderFactory = $tableBuilderFactory ?: ObjectManager::getInstance()
             ->get(BuilderInterfaceFactory::class);
+        $this->metadataPool = $metadataPool ?: ObjectManager::getInstance()->get(MetadataPool::class);
```
Comment (andrewbess) [inline app/code/Magento/Catalog/Model/Indexer/Product/Flat/TableBuilder.php]: Hello @SilinMykola 
Please use null-coalescing operator in constructor body for properties which can be null
```suggestion
        $this->tableBuilderFactory = $tableBuilderFactory ?? ObjectManager::getInstance()
            ->get(BuilderInterfaceFactory::class);
        $this->metadataPool = $metadataPool ?? ObjectManager::getInstance()->get(MetadataPool::class);
```

File: app/code/Magento/Catalog/Model/Product/Option/Type/Date.php [hunk 2]
```diff
@@ -340,7 +340,7 @@ public function is24hTimeFormat()
      */
     public function getYearStart()
     {
-        $_range = explode(',', $this->getConfigData('year_range'));
+        $_range = explode(',', $this->getConfigData('year_range') ?? '');
```
Comment (andrewbess) [inline app/code/Magento/Catalog/Model/Product/Option/Type/Date.php]: Hello @SilinMykola 
What do you think about suggestion below?

```suggestion
        $_range = $this->getConfigData('year_range') !== null ? explode(',', $this->getConfigData('year_range')) : [];
```

File: app/code/Magento/Catalog/Model/Product/Option/Type/Date.php [hunk 3]
```diff
@@ -355,7 +355,7 @@ public function getYearStart()
      */
     public function getYearEnd()
     {
-        $_range = explode(',', $this->getConfigData('year_range'));
+        $_range = explode(',', $this->getConfigData('year_range') ?? '');
```
Comment (andrewbess) [inline app/code/Magento/Catalog/Model/Product/Option/Type/Date.php]: Hello @SilinMykola 
What do you think about suggestion below?

```suggestion
        $_range = $this->getConfigData('year_range') !== null ? explode(',', $this->getConfigData('year_range')) : [];
```

File: app/code/Magento/Catalog/Model/Product/Option/Type/Text.php [hunk 4]
```diff
@@ -58,7 +58,7 @@ public function validateUserValue($values)
         parent::validateUserValue($values);
 
         $option = $this->getOption();
-        $value = trim($this->getUserValue());
+        $value = ($this->getUserValue() !== null) ? trim($this->getUserValue()) : '';
```
Comment (andrewbess) [inline app/code/Magento/Catalog/Model/Product/Option/Type/Text.php]: Hello @SilinMykola 
Please remove redundant brackets

```suggestion
        $value = $this->getUserValue() !== null ? trim($this->getUserValue()) : '';
```

File: app/code/Magento/Catalog/Ui/DataProvider/CatalogEavValidationRules.php [hunk 5]
```diff
@@ -33,14 +33,16 @@ public function build(ProductAttributeInterface $attribute, array $data)
             $rules['validate-zero-or-greater'] = true;
         }
 
-        $validationClasses = explode(' ', $attribute->getFrontendClass());
+        $validationClasses = explode(' ', $attribute->getFrontendClass() ?? '');
```
Comment (andrewbess) [inline app/code/Magento/Catalog/Ui/DataProvider/CatalogEavValidationRules.php]: Hello @SilinMykola 
What do you think about suggestion below

```suggestion
        $validationClasses = $attribute->getFrontendClass() ? explode(' ', $attribute->getFrontendClass()) : [];
```

File: app/code/Magento/CatalogImportExport/Model/Import/Product/Option.php [hunk 6]
```diff
@@ -1287,6 +1281,7 @@ protected function _importData()
                     $this->_rowProductId = $this->_productsSkuToId[$rowData[self::COLUMN_SKU]];
                     if (array_key_exists('custom_options', $rowData)
                         && (
+                            $rowData['custom_options'] === null ||
                             trim($rowData['custom_options']) === '' ||
                             trim($rowData['custom_options']) === $this->_productEntity->getEmptyAttributeValueConstant()
                         )
```
Comment (andrewbess) [inline app/code/Magento/CatalogImportExport/Model/Import/Product/Option.php:1287]: Hello @SilinMykola 
What do you think about suggestion below?

```suggestion
                    if (isset($rowData['custom_options'])
                        && (
                            trim($rowData['custom_options']) === '' ||
                            trim($rowData['custom_options']) === $this->_productEntity->getEmptyAttributeValueConstant()
                        )
```

File: app/code/Magento/Checkout/view/frontend/templates/cart/coupon.phtml [hunk 7]
```diff
@@ -9,14 +9,17 @@
  */
 
 // We should use strlen function because coupon code could be "0", converted to bool will lead to false
-$hasCouponCode = (bool) strlen($block->getCouponCode());
+$hasCouponCode = (bool) ($block->getCouponCode() ? strlen($block->getCouponCode()) : '');
```
Comment (andrewbess) [inline app/code/Magento/Checkout/view/frontend/templates/cart/coupon.phtml]: Hello @SilinMykola 
What do you think about suggestion below?

```suggestion
$hasCouponCode = $block->getCouponCode() !== null && strlen($block->getCouponCode() > 0);
```

File: app/code/Magento/CustomerImportExport/Model/Import/Address.php [hunk 8]
```diff
@@ -24,14 +24,12 @@
  */
 class Address extends AbstractCustomer
 {
-    /**#@+
-     * Attribute collection name
+    /**
+     * Customer Address EAV additional attribute collection name
```
Comment (andrewbess) [inline app/code/Magento/CustomerImportExport/Model/Import/Address.php]: I think description should be more informative

```suggestion
     * The Customer address attribute collection class name
```

File: app/code/Magento/Tax/Model/Calculation/RateRepository.php [hunk 9]
```diff
@@ -102,11 +102,14 @@ public function __construct(
         $this->regionFactory = $regionFactory;
         $this->resourceModel = $rateResource;
         $this->joinProcessor = $joinProcessor;
-        $this->collectionProcessor = $collectionProcessor ?: $this->getCollectionProcessor();
+        $this->collectionProcessor = $collectionProcessor ?: \Magento\Framework\App\ObjectManager::getInstance()->get(
+        // phpstan:ignore "Class Magento\Tax\Model\Api\SearchCriteria\TaxRateCollectionProcessor not found."
+            \Magento\Tax\Model\Api\SearchCriteria\TaxRateCollectionProcessor::class
+        );
```
Comment (andrewbess) [inline app/code/Magento/Tax/Model/Calculation/RateRepository.php]: Hello @SilinMykola 
Please use null-coalescing operator in constructor body for properties

```suggestion
        $this->collectionProcessor = $collectionProcessor ?? \Magento\Framework\App\ObjectManager::getInstance()->get(
        // phpstan:ignore "Class Magento\Tax\Model\Api\SearchCriteria\TaxRateCollectionProcessor not found."
            \Magento\Tax\Model\Api\SearchCriteria\TaxRateCollectionProcessor::class
        );
```

**Review by andrewbess — CHANGES_REQUESTED** [review]
Hello @SilinMykola 
Thank you for your contribution
Please check my comments
Thank you in advance

File: app/code/Magento/Catalog/Model/Indexer/Product/Flat/TableBuilder.php (hunk 1 above)
Comment (SilinMykola) [reply app/code/Magento/Catalog/Model/Indexer/Product/Flat/TableBuilder.php]: Hello @andrewbess 
Thanks for your review. I made changes, please check.

Comment (SilinMykola) [reply app/code/Magento/Catalog/Model/Indexer/Product/Flat/TableBuilder.php]: Hello @andrewbess 
Thanks for your review. I made changes, please check.

File: app/code/Magento/Catalog/Model/Product/Option/Type/Date.php (hunk 2 above)
Comment (SilinMykola) [reply app/code/Magento/Catalog/Model/Product/Option/Type/Date.php]: Hello @andrewbess 
Thanks for your review. I made changes, please check.

File: app/code/Magento/Catalog/Model/Product/Option/Type/Date.php (hunk 3 above)
Comment (SilinMykola) [reply app/code/Magento/Catalog/Model/Product/Option/Type/Date.php]: Hello @andrewbess 
Thanks for your review. I made changes, please check.

File: app/code/Magento/Catalog/Model/Product/Option/Type/Text.php (hunk 4 above)
Comment (SilinMykola) [reply app/code/Magento/Catalog/Model/Product/Option/Type/Text.php]: Hello @andrewbess 
Thanks for your review. I made changes, please check.

File: app/code/Magento/Catalog/Ui/DataProvider/CatalogEavValidationRules.php (hunk 5 above)
Comment (SilinMykola) [reply app/code/Magento/Catalog/Ui/DataProvider/CatalogEavValidationRules.php]: Hello @andrewbess 
Thanks for your review. I made changes, please check.

File: app/code/Magento/CatalogImportExport/Model/Import/Product/Option.php (hunk 6 above)
Comment (SilinMykola) [reply app/code/Magento/CatalogImportExport/Model/Import/Product/Option.php:1287]: Hello @andrewbess 
Thanks for your review. I made changes, please check.

File: app/code/Magento/Checkout/view/frontend/templates/cart/coupon.phtml (hunk 7 above)
Comment (SilinMykola) [reply app/code/Magento/Checkout/view/frontend/templates/cart/coupon.phtml]: Hello @andrewbess 
Thanks for your review. I made changes, please check.

File: app/code/Magento/Tax/Model/Calculation/RateRepository.php (hunk 9 above)
Comment (SilinMykola) [reply app/code/Magento/Tax/Model/Calculation/RateRepository.php]: Hello @andrewbess 
Thanks for your review. I made changes, please check.

**Comment by xmav** [comment]
@magento run all tests

**Comment by xmav** [comment]
@andrewbess Please check failed tests

**Comment by SilinMykola** [comment]
@magento run Static Tests

File: app/code/Magento/Tax/Model/Calculation/RateRepository.php [hunk 10]
```diff
@@ -270,7 +273,7 @@ private function validate(\Magento\Tax\Api\Data\TaxRateInterface $taxRate)
             );
         }
 
-        if (!\Zend_Validate::is(trim($taxRate->getCode()), 'NotEmpty')) {
+        if (!\Zend_Validate::is(trim($taxRate->getCode() ?? ''), 'NotEmpty')) {
```
Comment (xmav) [inline app/code/Magento/Tax/Model/Calculation/RateRepository.php]: as for me ` if ($taxRate->getCode() === null || !\Zend_Validate::is(trim($taxRate->getCode()), 'NotEmpty')) {` is better

File: app/code/Magento/Tax/Model/Calculation/RateRepository.php [hunk 11]
```diff
@@ -293,7 +296,7 @@ private function validate(\Magento\Tax\Api\Data\TaxRateInterface $taxRate)
                 $exception->addError(__('Range To should be equal or greater than Range From.'));
             }
         } else {
-            if (!\Zend_Validate::is(trim($taxRate->getTaxPostcode()), 'NotEmpty')) {
+            if (!\Zend_Validate::is(trim($taxRate->getTaxPostcode() ?? ''), 'NotEmpty')) {
```
Comment (xmav) [inline app/code/Magento/Tax/Model/Calculation/RateRepository.php]: same as above

File: app/code/Magento/Tax/Model/Calculation/RateRepository.php [hunk 12]
```diff
@@ -102,11 +102,14 @@ public function __construct(
         $this->regionFactory = $regionFactory;
         $this->resourceModel = $rateResource;
         $this->joinProcessor = $joinProcessor;
-        $this->collectionProcessor = $collectionProcessor ?: $this->getCollectionProcessor();
+        $this->collectionProcessor = $collectionProcessor ?? \Magento\Framework\App\ObjectManager::getInstance()->get(
+        // phpstan:ignore "Class Magento\Tax\Model\Api\SearchCriteria\TaxRateCollectionProcessor not found."
+            \Magento\Tax\Model\Api\SearchCriteria\TaxRateCollectionProcessor::class
```
Comment (xmav) [inline app/code/Magento/Tax/Model/Calculation/RateRepository.php]: lets use string value instead

File: app/code/Magento/Tax/Model/Calculation/Rule/Validator.php [hunk 13]
```diff
@@ -73,7 +73,7 @@ public function isValid($value)
         }
 
         // Code is required
-        if (!\Zend_Validate::is(trim($value->getCode()), 'NotEmpty')) {
+        if (!\Zend_Validate::is(trim($value->getCode() ?? ''), 'NotEmpty')) {
```
Comment (xmav) [inline app/code/Magento/Tax/Model/Calculation/Rule/Validator.php]: `if ($taxRate->getCode() === null || !\Zend_Validate::is(trim($taxRate->getCode()), 'NotEmpty')) { ` could be a better option

File: setup/src/Magento/Setup/Test/Unit/Module/Setup/SetupCacheTest.php [hunk 14]
```diff
@@ -60,7 +60,7 @@ public function testGetNonexistent($field)
     public function getNonexistentDataProvider()
     {
         return [
-            [null],
+            [''],
```
Comment (xmav) [inline setup/src/Magento/Setup/Test/Unit/Module/Setup/SetupCacheTest.php]: What is the reason to change that ?

File: app/code/Magento/Catalog/Ui/DataProvider/CatalogEavValidationRules.php [hunk 15]
```diff
@@ -33,14 +33,17 @@ public function build(ProductAttributeInterface $attribute, array $data)
             $rules['validate-zero-or-greater'] = true;
         }
 
-        $validationClasses = explode(' ', $attribute->getFrontendClass());
+        $validationClasses = $attribute->getFrontendClass()
+            ? explode(' ', $attribute->getFrontendClass()) : [];
 
         foreach ($validationClasses as $class) {
             if (preg_match('/^maximum-length-(\d+)$/', $class, $matches)) {
+                // phpcs:ignore Magento2.Performance.ForeachArrayMerge
```
Comment (xmav) [inline app/code/Magento/Catalog/Ui/DataProvider/CatalogEavValidationRules.php:42]: Could it be fixed ?
https://github.com/magento/magento-coding-standard/blob/develop/Magento2/Sniffs/Performance/ForeachArrayMergeSniff.md

File: app/code/Magento/Catalog/Model/Product/Option/Type/Date.php [hunk 16]
```diff
@@ -355,7 +356,8 @@ public function getYearStart()
      */
     public function getYearEnd()
     {
-        $_range = explode(',', $this->getConfigData('year_range'));
+        $_range = $this->getConfigData('year_range')
```
Comment (xmav) [inline app/code/Magento/Catalog/Model/Product/Option/Type/Date.php]: please use same approach\code style in both cases

**Review by xmav — CHANGES_REQUESTED** [review]

File: setup/src/Magento/Setup/Test/Unit/Module/Setup/SetupCacheTest.php (hunk 14 above)
Comment (SilinMykola) [reply setup/src/Magento/Setup/Test/Unit/Module/Setup/SetupCacheTest.php]: @xmav Hello! Thanks for your review, I reverted this change.

File: app/code/Magento/Tax/Model/Calculation/Rule/Validator.php (hunk 13 above)
Comment (SilinMykola) [reply app/code/Magento/Tax/Model/Calculation/Rule/Validator.php]: @xmav I changed, thanks for your help!

File: app/code/Magento/Tax/Model/Calculation/RateRepository.php (hunk 12 above)
Comment (SilinMykola) [reply app/code/Magento/Tax/Model/Calculation/RateRepository.php]: @xmav I changed, thanks for your help!

File: app/code/Magento/Tax/Model/Calculation/RateRepository.php (hunk 11 above)
Comment (SilinMykola) [reply app/code/Magento/Tax/Model/Calculation/RateRepository.php]: @xmav I changed, thanks for your help!

File: app/code/Magento/Tax/Model/Calculation/RateRepository.php (hunk 10 above)
Comment (SilinMykola) [reply app/code/Magento/Tax/Model/Calculation/RateRepository.php]: @xmav I changed, thanks for your help!

File: app/code/Magento/Catalog/Ui/DataProvider/CatalogEavValidationRules.php (hunk 15 above)
Comment (SilinMykola) [reply app/code/Magento/Catalog/Ui/DataProvider/CatalogEavValidationRules.php:42]: @xmav hello! Thanks for your review. Please, check my changes.
Thanks!

File: app/code/Magento/Catalog/Model/Product/Option/Type/Date.php (hunk 16 above)
Comment (SilinMykola) [reply app/code/Magento/Catalog/Model/Product/Option/Type/Date.php]: I made changes, please, check.
Thanks!

**Comment by SilinMykola** [comment]
@magento run all tests

**Comment by andrewbess** [comment]
@magento run all tests

**Review by xmav — APPROVED** [review]

**Comment by andrewbess** [comment]
@magento run Magento Health Index

**Comment by andrewbess** [comment]
@magento run Magento Health Index

**Comment by andrewbess** [comment]
@magento run all tests

**Comment by karyna-tsymbal-atwix** [comment]
@magento run Static Tests

**Comment by xmav** [comment]
@magento run all tests

**Comment by xmav** [comment]
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B

**Comment by xmav** [comment]
Please check functional tests failures. They does not seem to be random

**Comment by karyna-tsymbal-atwix** [comment]
@magento run Functional Tests CE

**Comment by andrewbess** [comment]
@magento run all tests

File: app/code/Magento/CatalogImportExport/Model/Import/Product/Option.php [hunk 17]
```diff
@@ -2098,6 +2093,7 @@ protected function _parseCustomOptions($rowData)
             $optionValueParams = explode($this->_productEntity->getMultipleValueSeparator(), $optionValue);
             foreach ($optionValueParams as $nameAndValue) {
                 $nameAndValue = explode('=', $nameAndValue);
+                // @phpstan-ignore-next-line
```
Comment (xmav) [inline app/code/Magento/CatalogImportExport/Model/Import/Product/Option.php]: Static test failure seems to be valid. call to `explode` with empty string returns array with single element( which is not empty). Is there a case when explode will return "empty" value ?
Could you please do corresponding changes ?

Comment (andrewbess) [reply app/code/Magento/CatalogImportExport/Model/Import/Product/Option.php]: Hello @xmav 
I think we need to refactor it with additional test coverage
We are going to do it tomorrow

File: app/code/Magento/Catalog/Model/Product/Option/Type/Date.php [hunk 18]
```diff
@@ -302,7 +305,10 @@ public function prepareOptionValueForRequest($optionValue)
         $infoBuyRequest = $confItem->getOptionByCode('info_buyRequest');
         try {
             $value = $this->serializer->unserialize($infoBuyRequest->getValue());
-            if (is_array($value) && isset($value['options']) && isset($value['options'][$this->getOption()->getId()])
+
+            if (is_array($value)
+                && isset($value['options'])
+                && isset($value['options'][$this->getOption()->getId()])
             ) {
```
Comment (andrewbess) [inline app/code/Magento/Catalog/Model/Product/Option/Type/Date.php]: Hello @karyna-tsymbal-atwix 
Please fix this condition like below

```suggestion
            if (is_array($value) && isset($value['options'][$this->getOption()->getId()])) {
```

**Review by andrewbess — CHANGES_REQUESTED** [review]
Hello @karyna-tsymbal-atwix 
Could you please check my comments?
Thank you in advance.

File: app/code/Magento/Catalog/Model/Product/Option/Type/Date.php (hunk 18 above)
Comment (karyna-t) [reply app/code/Magento/Catalog/Model/Product/Option/Type/Date.php]: hi @andrewbess , done

File: app/code/Magento/CatalogImportExport/Model/Import/Product/Option.php (hunk 17 above)
Comment (karyna-t) [reply app/code/Magento/CatalogImportExport/Model/Import/Product/Option.php]: @xmav I removed the unnecessary 'if'.
@andrewbess  as discussed, the complete refactoring with a new test cases will be implemented in the future due to complexity.

**Comment by karyna-tsymbal-atwix** [comment]
@magento run all tests

**Review by andrewbess — APPROVED** [review]
Hello @SilinMykola @karyna-tsymbal-atwix 
Thank you for your work here
Approved ✔️

**Comment by xmav** [comment]
pending delivery https://github.com/magento-commerce/magento2ce/pull/7295

**Pull request merged**